### ISSUE TITLE
LIONsolver abstract class

### ISSUE THREAD

**Issue by AnderBiguri** [CambridgeCIA/LION#46]
a class to contain abstractions for a ML solver. 

Attributes:

- [ ] optimizer parameters/info #45  
- [ ] DataLoaders (input params)
- [ ] Loop params (e.g. `n_iter`)
- [ ] Validation parameters
- [ ] Loss
- [ ] Metadata (all anything else required)

Methods:

- [ ] `step(DataLoader, train/validation) # applies a minibatch DataLoader iteration`
- [ ] `train_step() # just step() with certain inputs` 
- [ ] `validation_step() # just step() with certain inputs`
- [ ] `is_validation_time(i) # or similar name, True if current i is a validation itertion, according to the params set`
- [ ] `train_epoch() # applies full DataLoader epoch`
- [ ] `validation_epoch() # applies full Valiadtion epoch`
- [ ] `init_from_params() # makes an instance with settings from a json file`

Still unsure, but I think testing should also be here
- [ ] `test()`
